ui/src/app/views/workspace-share/component.ts

```diff
@@ -61,32 +69,72 @@ export class WorkspaceShareComponent implements OnInit {
     }
   }
 
+  convertToEmail(username: string): string {
+    return username + '@fake-research-aou.org';
+  }
+
+  inputChange(): void {
+    this.userNotFound = false;
+  }
+
   addCollaborator(): void {
-    this.workspace.userRoles.push({email: this.toShare, role: this.accessLevel});
-    this.workspacesService.shareWorkspace(
-      this.workspace.namespace, this.workspace.id, {
-        workspaceEtag: this.workspace.etag,
-        items: this.workspace.userRoles
-      }).subscribe((resp: ShareWorkspaceResponse) => {
-        this.workspace.etag = resp.workspaceEtag;
+    if (this.inputNotInitialized) {
+      this.inputNotInitialized = false;
+      this.input.nativeElement.addEventListener('keydown', () => {
+        this.userNotFound = false;
       });
+    }
+    this.usersLoading = true;
+    this.updateList = Array.from(this.workspace.userRoles);
+    this.updateList.push({email: this.convertToEmail(this.toShare),
+        role: this.accessLevel});
+
+    this.errorHandlingService.retryApi(
+      this.workspacesService.shareWorkspace(
+        this.workspace.namespace,
+        this.workspace.id, {
+          workspaceEtag: this.workspace.etag,
+          items: this.updateList})).subscribe(
+      (resp: ShareWorkspaceResponse) => {
+        this.workspace.etag = resp.workspaceEtag;
+        this.usersLoading = false;
+        this.workspace.userRoles = this.updateList;
```

Comment:
Seems like this could get hairy if there are concurrent pending requests - `this.updateList` may have changed since you sent it above. Perhaps `updateList` here should be local? Though there are likely many ways to resolve this e.g. disallowing concurrent share requests in some way or else chain the executions of these add/removes.

It should actually always fail to send concurrent requests, since the etag would be stale for one of them, so disabling or queueing these actions probably makes sense (if queueing - lazily reading the current etag before the share request).

Reply:
`this.updateList` should be re-created every time `addCollaborator` or `removeCollaborator` is called, so yes, it should be local! Will change. I'm doing it with a separate array to avoid updating the screen until after the API call has actually succeeded. If there are two concurrent requests, one will error with a 409, right? In that case it should go to the error response, rather than the success response. I can work on adding an error case for conflict errors.

Reply:
> disallowing concurrent share requests

> If there are two concurrent requests, one will error with a 409

Better to just cancel any existing outbound requests that have not yet returned (I'd probably use a `switchMap`).  Less work for the servers and less error handling for you to do.